### PR TITLE
improve stoploss validation

### PR DESCRIPTION
**Pull request by xmatthias**
# Summary
Stoploss with 0 is problematic, since that would lead to sells immediately.

This should prevent the bog from starting up - otherwise it results in misleading bug reports (as #2171).

Solve the issue: #2171

## Quick changelog

- Validate stoploss_positive !=0 (it can be positive or negative, since we use `-1 * abs()` when applying this).
- validate stoploss != 0 (same `-1 * abs()` logic as before

### CODE REVIEW

**Comment by hroff-1902** [comment]
I think it won't keep me from getting those errors from exchange, since I would set stoploss = -0.00001 in order to circumvent this artificial limitation. And I'll get same error from exchange.
Instead, those messages from exchange should be handled and the pair sold immediately. Or an appropriate flag for the order (yes, exchange specific, but we have exchange classes for it) should be set to force it to sell the pair immediately in these cases...

**Comment by xmatthias** [comment]
If i want to circumvent this limitation, then i'll comment out the calls to this check-method.

It's about **accidentally** setting this wrongly, which is what happened in #2171.

If i want to circumvent this with -0.0000001 - then i'm an idiot and should not be using the bot since i'm trying to create an error.

Yes, the error can be thrown anyway (see #1557 / https://github.com/freqtrade/freqtrade/issues/1557#issuecomment-524165920) - but that's not because of the value - it can be also thrown if i set it to -0.05, which is something we cannot protect from.

**Comment by hroff-1902** [comment]
> If i want to circumvent this limitation, then i'll comment out the calls to this check-method.

@xmatthias you are cool 🍺 , I have not meant it. It's a user, not you, who can find this limitation (stoploss=0.0 is forbidden) and circumvent it (aha, stoploss=0.00001 is okay!)

> It's about **accidentally** setting this wrongly, which is what happened in #2171.

I think it's not wrong. It also has meaning -- close the trade immediately (well, in a few seconds, probably) if the price stays same or move a small fraction down. That's the meaning of this stoploss. As any other small stoplosses, like 0.00001. Nothing is more errorneous in stoploss=0.0 than in stoploss=0.000001. All of them are equally wrong or equally can be assigned some (right, stupid) meaning. And you won't be able to figure out the border between those accidentally wrong stoploss values and the reasonable ones. The only way is to handle them *without exchange errors*, even if those errors are exchange-specific. That means they are to be handled in the exchange-specific manner.

> If i want to circumvent this with -0.0000001 - then i'm an idiot and should not be using the bot since i'm trying to create an error.

Many people behave strange and even like idiots and many of them have their own intrinsic reasons and explanation for this... The bot should behave according to the settings defined. As I said above, small stoplosses can have they own (probably useless) meaning and corresponding expected bot behavior.

> Yes, the error can be thrown anyway (see #1557 / [#1557 (comment)](https://github.com/freqtrade/freqtrade/issues/1557#issuecomment-524165920)) - but that's not because of the value - it can be also thrown if i set it to -0.05, which is something we cannot protect from.

Well, let's commit it, this limitation on the stoploss values, no problem. But it does not resolve the actual problem (stoploss is not set on exchange, the order is not created (or a buy order is created, but without a protective stoploss? Which is even worse in this case...)). And, imho, when the real problem will be fixed (in the exchange-specific way, probably), it may happen that this limit was not actually needed, because this particular edge case will be handled on exactly same manner as other small stoplosses...

File: freqtrade/configuration/config_validation.py [hunk 1]
```diff
@@ -79,13 +83,20 @@ def _validate_trailing_stoploss(conf: Dict[str, Any]) -> None:
...
         if tsl_positive == 0.0:
             raise OperationalException(
-                f'The config trailing_only_offset_is_reached needs '
+                'The config trailing_only_offset_is_reached needs '
                 'trailing_stop_positive_offset to be more than 0 in your config.')
     if tsl_positive > 0 and 0 < tsl_offset <= tsl_positive:
         raise OperationalException(
-            f'The config trailing_stop_positive_offset needs '
+            'The config trailing_stop_positive_offset needs '
             'to be greater than trailing_stop_positive_offset in your config.')
```
Comment (hroff-1902) [inline freqtrade/configuration/config_validation.py]: Emm, "trailing_stop_offset ... than trailing_stop_positive"?

Comment (xmatthias) [reply freqtrade/configuration/config_validation.py]: Great catch :+1:

**Comment by xmatthias** [comment]
> But it does not resolve the actual problem 

i never said it will ... I hope it will prevent accidentally configuring it wrongly (which is the same for all checks we do in config validation ...).

> it may happen that this limit was not actually needed, because this particular edge case will be handled on exactly same manner as other small stoplosses...

You're right, 0.00001 is just as wrong as 0.0, but it's difficult to define a meaningfull limit.

**Pull request merged**